src/main/java/org/elasticsearch/transport/TransportService.java

```diff
@@ -209,18 +226,27 @@ public BoundTransportAddress boundAddress() {
     }
 
     public boolean nodeConnected(DiscoveryNode node) {
-        return transport.nodeConnected(node);
+        return node.id().equals(localNodeId) || transport.nodeConnected(node);
     }
 
     public void connectToNode(DiscoveryNode node) throws ConnectTransportException {
+        if (node.id().equals(localNodeId)) {
+            return;
+        }
         transport.connectToNode(node);
     }
 
     public void connectToNodeLight(DiscoveryNode node) throws ConnectTransportException {
+        if (node.id().equals(localNodeId)) {
+            return;
+        }
         transport.connectToNodeLight(node);
     }
 
     public void disconnectFromNode(DiscoveryNode node) {
+        if (node.id().equals(localNodeId) && localNodeId != null) {
```

Comment:
the second non null check is not needed? also, after connectToLight we might disconnect (like someone listing the local node in the hosts lists for unicast discovery), should we really fail? If connect is noop, maybe this needs to be noop for local node?